controller/array_action/array_mediator_svc.py

```diff
@@ -1375,3 +1397,86 @@ def _add_snapshot(self, snapshot_name, source_cli_volume, pool):
+                if NAME_NOT_EXIST_OR_MEET_RULES in ex.my_message:
+                    raise array_errors.PoolDoesNotExist(pool, self.endpoint)
+                if NOT_ENOUGH_EXTENTS_IN_POOL_CREATE in ex.my_message:
+                    raise array_errors.NotEnoughSpaceInPool(id_or_name=pool)
+                if any(msg_id in ex.my_message for msg_id in (NON_ASCII_CHARS, INVALID_NAME, TOO_MANY_CHARS)):
+                    raise array_errors.IllegalObjectName(ex.my_message)
+                raise ex
+        return None
+
+    def _get_volume_id_from_volume_group(self, volume_group_id):
+        filter_value = 'volume_group_id={}'.format(volume_group_id)
+        cli_volume = self.client.svcinfo.lsvdisk(bytes=True, filtervalue=filter_value).as_single_element
+        return cli_volume.id
+
+    def _rollback_create_volume_from_snapshot(self, volume_id):
+        self._delete_volume_by_name(volume_id)
+
+    def _remove_from_volume_group(self, volume_id):
+        self._chvdisk(vdisk_id=volume_id, novolumegroup=True)
+
+    def _rename_volume(self, volume_id, name):
+        try:
+            self._chvdisk(vdisk_id=volume_id, name=name)
+        except array_errors.VolumeAlreadyExists as ex:
+            self._rollback_create_volume_from_snapshot(volume_id)
+            raise ex
+
+    def _chvdisk(self, **kwargs):
+        try:
+            self.client.svctask.chvdisk(**kwargs)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if is_warning_message(ex.my_message):
+                logger.warning(
+                    "exception encountered while changing volume parameters '{}': {}".format(kwargs, ex.my_message))
+            else:
+                if OBJ_ALREADY_EXIST in ex.my_message:
+                    raise array_errors.VolumeAlreadyExists(kwargs, self.endpoint)
+                raise ex
+
+    def _rmvolumegroup(self, volume_group_id):
+        logger.info("deleting volume group : {0}".format(volume_group_id))
+        try:
+            self.client.svctask.rmvolumegroup(object_id=volume_group_id)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if is_warning_message(ex.my_message):
+                logger.warning("exception encountered during deletion of volume group {}: {}".format(volume_group_id,
+                                                                                                     ex.my_message))
+            else:
+                logger.error("Failed to delete volume group {}".format(volume_group_id))
+                if OBJ_NOT_FOUND in ex.my_message or VOL_NOT_FOUND in ex.my_message:
+                    logger.warning(array_errors.ObjectNotFoundError(volume_group_id))
+                raise ex
+
+    def _extract_and_remove_volume_group(self, volume_id, volume_group_id):
+        self._remove_from_volume_group(volume_id)
+        self._rmvolumegroup(volume_group_id)
+
+    def _is_source_have_fcmaps(self, object_uid):
```

Comment:
```suggestion
    def _is_source_has_fcmaps(self, object_uid):
```
neither is english, but still, this sounds better to me

Reply:
done
any better name?

Reply:
nope